### PR TITLE
OpInfo: where

### PR DESCRIPTION
**Pull request by kshitij12345**
Reference: #54261

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -3550,14 +3550,35 @@ def sample_inputs_expand_as(op_info, device, dtype, requires_grad, **kwargs):
...
 
 
+def sample_inputs_where(op_info, device, dtype, requires_grad, **kwargs):
+    make_arg = partial(make_tensor, dtype=dtype, device=device, requires_grad=requires_grad)
+    make_bool_mask = partial(make_tensor, dtype=torch.bool, device=device, requires_grad=requires_grad)
+
+    WhereCases = collections.namedtuple("WhereCases",
+                                        ["shape", "mask_shape", "other_shape", "broadcasts_input"])
+
+    cases = (WhereCases(shape=(M, M), mask_shape=(M, M), other_shape=(M, M), broadcasts_input=False),
```
Comment (kshitij12345) [inline torch/testing/_internal/common_methods_invocations.py:3567]: Too verbose? Or easy to read?

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:3567]: It is easy to read but at the point this is written why not just do:

```
yield SampleInput(make_arg(M, M), args=(make_bool_mask(M, M), make_arg(M, M)))
```

instead of each case?

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -6218,6 +6239,14 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
                    dtypesIfCUDA=all_types_and(torch.bool, torch.half, torch.bfloat16),
                    sample_inputs_func=sample_inputs_logit,
                    safe_casts_outputs=True),
+    OpInfo('where',
+           op=lambda self, condition, other: torch.where(condition, self, other),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:6243]: This is going to need a comment; this is because we don't support the input being anything but the first argument, right?

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -6218,6 +6239,14 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
...
+           sample_inputs_func=sample_inputs_where,
+           supports_out=False,
+           skips=(
+               SkipInfo('TestCommon', 'test_variant_consistency_jit'),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:6247]: Skip will need a comment, too

**Review by mruberry — COMMENTED** [review]
Looks good as usual @kshitij12345, I made a few comments inline for your review

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:6243]: Yes. For this specific case we can pass input as condition but since condition is always bool, so we won't gradcheck on it.

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:3567]:
```python
cases = (((M, M), (M, M), (M, M), False),
             ((M, 1, M), (M, M), (M, M, 1), True),
             ((), (), (), False),
             ((M, 1, M), (), (M, M, 1), True),
             ((), (M, M), (), True),)

def generator():
    for shape, mask_shape, other_shape, broadcasts_input in cases:
            yield SampleInput(make_arg(shape),
                              args=(make_bool_mask(mask_shape), make_arg(other_shape)),
                              broadcasts_input=broadcasts_input)
```
Is it ok to stick to usual `cases` as tuple. Even `yield` for every sample input is very verbose and hard to extend the cases if need be (ever).

**Comment by kshitij12345** [comment]
@mruberry gentle ping :)

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -6523,11 +6553,6 @@ def method_tests():
         ('resize_as_', (), (non_differentiable(torch.tensor(5.)),), 'scalar'),
         ('resize_as_', (), (non_differentiable(torch.randn((1, 1, 1))),), 'scalar_to_dims'),
         ('resize_as_', (S, S, S), (non_differentiable(torch.randn(S * S, S)),)),
-        ('where', (M, M), (mask_not_all_zeros((M, M)), (M, M)), '', (True,)),
-        ('where', (M, 1, M), (mask_not_all_zeros((M, M)), (M, M, 1)), 'broadcast_all', (True,)),
-        ('where', (), (bernoulli_scalar(), ()), 'scalar', (True,)),
-        ('where', (M, 1, M), (bernoulli_scalar(), (M, M, 1)), 'scalar_broadcast_mask', (True,)),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:6529]: Follow-up TODO: delete the bernoulli_scalar() function, it's only used in one sample input and can be replaced with make_tensor(). Also I think it's incorrectly implemented?

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -3553,14 +3553,32 @@ def sample_inputs_expand_as(op_info, device, dtype, requires_grad, **kwargs):
...
+        for shape, shape_other in cases:
             yield(SampleInput(make_arg(shape, requires_grad=requires_grad),
                               args=(make_arg(shape_other, requires_grad=False), )))
 
     return list(generator())
 
 
+def sample_inputs_where(op_info, device, dtype, requires_grad, **kwargs):
+    make_arg = partial(make_tensor, dtype=dtype, device=device, requires_grad=requires_grad)
+    make_bool_mask = partial(make_tensor, dtype=torch.bool, device=device, requires_grad=requires_grad)
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:3565]: The function this is replacing guaranteed that at least one value in the mask is True

This PR should preserve that behavior. It can do so by creating its own make_bool_mask function that checks if the tensor contains at least one True value and, if not, inserts a True a "True" at a random point, for example.

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:6270]: Can we use gradcheck_wrapper to handle the gradcheck case? Or are there other tests which would still be an issue?

**Review by mruberry — COMMENTED** [review]
Overall looks good, @kshitij12345; I made a few inline comments for your review

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:6270]: Also affect `test_variant_consistency_eager`
https://github.com/pytorch/pytorch/blob/28840b9a447e820bf2e66f6b6cdc880c12d077b7/test/test_ops.py#L303-L307

File: torch/testing/_internal/common_methods_invocations.py (hunk 4 above)
Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:6529]: Why incorrectly?
```python
>>> torch.tensor(0, dtype=torch.bool).bernoulli_()
tensor(True)
>>> torch.tensor(0, dtype=torch.bool).bernoulli_()
tensor(False)
>>> torch.tensor(0, dtype=torch.bool).bernoulli_()
tensor(False)
>>> torch.tensor(0, dtype=torch.bool).bernoulli_()
tensor(False)
>>> torch.tensor(0, dtype=torch.bool).bernoulli_()
tensor(True)
>>> torch.tensor(0, dtype=torch.bool).bernoulli_()
tensor(False)
```

File: torch/testing/_internal/common_methods_invocations.py (hunk 5 above)
Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:3565]: Is it okay to use `mask_not_all_zero`?

File: torch/testing/_internal/common_methods_invocations.py (hunk 4 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:6529]: Aha. You're right, because torch.bernoulli() and torch.tensor.bernoulli_() are actually different functions, which is a frustrating UX consistency

Actually, would you file an issue for this? 

Docs for bernoulli_():

https://pytorch.org/docs/master/generated/torch.Tensor.bernoulli_.html?highlight=bernoulli_#torch.Tensor.bernoulli_

Docs for bernoulli():

https://pytorch.org/docs/master/generated/torch.bernoulli.html?highlight=bernoulli#torch.bernoulli

The function takes the likelihood from the tensor while the inplace variant takes the input from its `p` input

File: torch/testing/_internal/common_methods_invocations.py (hunk 5 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:3565]: Sure; although eventually it'd be good to improve upon it and not just redraw

Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:3565]: @mruberry take a look at the updated comment. I am confused.

File: torch/testing/_internal/common_methods_invocations.py (hunk 4 above)
Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:6529]: Sure!

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:6270]: OK

File: torch/testing/_internal/common_methods_invocations.py (hunk 5 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:3565]: I think some of the indexing ops do allow the indices (and maybe masks) to be on the CPU:

```
import torch

t = torch.tensor((1, 2, 3), device='cuda')
indices = torch.tensor((1,))
values = torch.tensor((4,), device='cuda')
t.index_put_((indices,), values)
: tensor([1, 4, 3], device='cuda:0')
```

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:3565]: In the code it looks like the function first manipulates the tensors according to the indices argument, then calls TensorIterator on the input and values tensors only.

Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:3565]: Is this behaviour intended?

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:3565]: I think so; and it would be BC-breaking to change

Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:3565]: How does this look?
```python
    def make_bool_mask(shape):
        # Make sure atleast one element is nonzero,
        # except for empty tensor
        mask_t = make_tensor(shape, dtype=torch.bool, device=device, requires_grad=False)
        if mask_t.numel() == 1 and len(mask_t.shape) > 0:
            mask_t[0] = True
            return mask_t
        elif reduce(lambda x, y: x*y, mask_t.shape, 1) == 0:
            return mask_t

        def random_index(shape):
            return tuple(map(lambda max_idx: random.randint(0, max_idx), shape))

        mask_t[random_index(mask_t.shape)] = True
        return mask_t
```

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:3565]: The handling of tensors with zero elements looks good but I think that check can be simplified to `mask_t` having numel == 0? The handling of the general case also looks correct.

I think the tricky case, then, is if shape is `()` or `(1,)`. For these I think you need two checks, one for numel() == 1 and a then a second check for len(shape) == 0. In both these cases the initial construction of the tensor is irrelevant, because you know the tensor you want to return, so I would put these checks before the make_tensor case, short-circuit on 0 numel, and then have the random_index case guarded behind the call to sum()

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:3565]:
```
if numel == 1:
  if len(shape) == 0:
    return torch.tensor(True, ...)
  return torch.Tensor((True,), ...)

mask = make_tensor(...)
if mask.numel() == 0:
  return mask

if mask.sum() == 0:
  # sets a random index to True

return mask
```

Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:3565]: Tried to incorporate the suggestion
How about?
```python

    def make_bool_mask(shape):
        # Make sure atleast one element is nonzero,
        # except for empty tensor
        mask_t = make_tensor(shape, dtype=torch.bool, device=device, requires_grad=False)

        if mask_t.numel() == 0:
            return mask_t
        elif mask_t.numel() == 1:
            # Added these after make tensor so that
            # we have to specify `dtype`, `device`, etc only
            # once to make_tensor
            mask_t.fill_(True)
            return mask_t

        if mask_t.sum() == 0:
            def random_index(shape):
                return tuple(map(lambda max_idx: random.randint(0, max_idx), shape))

            mask_t[random_index(mask_t.shape)] = True
            return mask_t

        return mask_t

```

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:3565]: LGTM!

**Comment by kshitij12345** [comment]
Note to self: look/fix failing NNC test.

**Comment by mruberry** [comment]
> Note to self: look/fix failing NNC test.

I think we just need to add this op to the list stated in the test. @Chillee, seems like we should put this metadata into the OpInfo so people don't have to find yet another list?

**Comment by kshitij12345** [comment]
@mruberry gentle ping :) 
Failures look irrelevant

**Review by mruberry — APPROVED** [review]
Nice work @kshitij12345! Thanks for the ping
